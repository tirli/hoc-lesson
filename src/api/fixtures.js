export default {
  "Search": [
    {
      "Title": "The Test",
      "Year": "2013",
      "imdbID": "tt2616114",
      "Type": "movie",
      "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMjMzMDQwMzM2M15BMl5BanBnXkFtZTcwMzA1OTg1OQ@@._V1_SX300.jpg"
    },
    {
      "Title": "Sound Test for Blackmail",
      "Year": "1929",
      "imdbID": "tt0249159",
      "Type": "movie",
      "Poster": "N/A"
    }
  ],
  "totalResults": "618",
  "Response": "True"
}