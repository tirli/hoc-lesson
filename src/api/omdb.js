const baseUrl = 'http://www.omdbapi.com/?';

export function search(title) {
  return fetch(`${baseUrl}s=${title}&r=json`)
    .then((res) => res.json())
    .then((result) => result.Search.map((item) => ({
      ...item,
      Poster: item.Poster === 'N/A' ? null : item.Poster,
    })))
  ;
}

export function getById(id, plotType = 'short') {
  return fetch(`${baseUrl}i=${id}&r=json&plot=${plotType}`)
    .then((res) => res.json());
};
