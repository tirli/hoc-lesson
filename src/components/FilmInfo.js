import React, { Component } from 'react';

import { getById } from '../api/omdb';

export default class FilmInfo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      item: null,
      loading: false,
    }
  }
  componentWillReceiveProps(nextProps) {
    this.setState({ loading: true });

    getById(nextProps.itemId).then((result) => this.setState({
      item: result,
      loading: false,
    }));
  }

  render() {
    const { item, loading } = this.state;
    if (loading) {
      return <div className="spinner"> ... Loading </div>
    }
    if (!item) {
      return <div> Please select any film </div>
    }

    return (
      <div>
        <img src={item.Poster}  />
        <div>
          <h1> {item.Title} </h1>
          <span> {item.Actors} </span>
          <span> {item.Genre} </span>
        </div>
      </div>
    );
  }
}