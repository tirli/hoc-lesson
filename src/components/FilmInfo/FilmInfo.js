import React, { Component } from 'react';
import withState from 'recompose/withState';
import compose from 'recompose/compose';

import { getById } from '../../api/omdb';
import windowWidth from '../../helpers/windowWidth';
import loading from '../../helpers/loading';
import perfomanceTest from '../../helpers/perfomanceTest';

export class FilmInfo extends Component {
  componentWillReceiveProps(nextProps) {
    const { windowWidth, itemId, handleLoading, handleItem } = nextProps;

    if (itemId && this.props.itemId !== itemId) {
      handleLoading(true);
      const plotSize = windowWidth > 768 ? 'full' : 'short';

      return getById(itemId, plotSize)
        .then((result) => {
          handleLoading(false);
          handleItem(result);
        })
        .catch(e => {
          handleLoading(false);
        })
      ;
    }
  }

  render() {
    const { item } = this.props;

    if (!item) {
      return null;
    }

    return (
      <div>
        <img src={item.Poster} alt={item.Title} />
        <div>
          <h1> {item.Title} </h1>
          <p> Actors: {item.Actors} </p>
          <p> Genres: {item.Genre} </p>
          <p> {item.Plot} </p>
        </div>
      </div>
    );
  }
}

const withStateLoading = withState('loading', 'handleLoading', false);
const withStateItem = withState('item', 'handleItem', null);
const loadingPredicate = (props) => props.loading;
export default compose(
  windowWidth,
  withStateLoading,
  withStateItem,
  loading(loadingPredicate),
  perfomanceTest,
)(FilmInfo);
