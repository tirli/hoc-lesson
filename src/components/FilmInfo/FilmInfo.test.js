import React from 'react';
import ReactDOM from 'react-dom';
import { FilmInfo } from './FilmInfo';
import { shallow, mount } from 'enzyme';
import ReactTestUtils from 'react-addons-test-utils'

import * as api from '../../api/omdb';

describe('FilmInfo', () => {
  it('shallow renders as expected', () => {
    const filmInfo = shallow(<FilmInfo />).html();
    expect(filmInfo).toMatchSnapshot();
  });

  it('componentWillReceiveProps should make request when itemId changed', async () => {
    const p = Promise.resolve('test result')
    api.getById = jest.fn(() => p);
    const handleLoading = jest.fn();
    const handleItem = jest.fn();
    const app = mount(<FilmInfo windowWidth={1920} itemId={null} />);
    const newProps = { itemId: 12, handleLoading, handleItem };

    await app.instance().componentWillReceiveProps(newProps);

    expect(api.getById).toHaveBeenCalledTimes(1);
    expect(handleLoading).toHaveBeenCalledTimes(2);
    expect(handleLoading).toHaveBeenCalledWith(true);
    expect(handleLoading).lastCalledWith(false);
    expect(handleItem).toHaveBeenCalled();
    expect(handleItem).toHaveBeenCalledWith('test result');
  })

  it('componentWillReceiveProps should not do anything if itemId is not changed', async () => {
    api.getById = jest.fn(() => Promise.resolve('test result'));
    const handleLoading = jest.fn();
    const handleItem = jest.fn();
    const app = mount(<FilmInfo windowWidth={1920} itemId={3} />);
    const newProps = { itemId: 3, handleLoading, handleItem };

    await app.instance().componentWillReceiveProps(newProps);

    expect(api.getById).toHaveBeenCalledTimes(0);
    expect(handleLoading).toHaveBeenCalledTimes(0);
    expect(handleItem).toHaveBeenCalledTimes(0);
  })

})

