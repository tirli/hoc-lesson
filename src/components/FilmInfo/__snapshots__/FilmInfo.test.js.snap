exports[`FilmInfo shallow renders as expected 1`] = `null`;
